File: BrainExtraction/ThreeTissue/train_model.py
import ants
import antspynet

import glob
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1"
os.environ["ITK_DEFAULT_GLOBAL_NUMBER_OF_THREADS"] = "4"

import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K

from batch_generator import batch_generator

K.clear_session()

base_directory = '/home/ntustison/Data/BrainExtraction/'
scripts_directory = base_directory + 'Scripts/'

from batch_generator import batch_generator

template = ants.image_read(antspynet.get_antsxnet_data("nki"))
template_size = template.shape

################################################
#
#  Create the model and load weights
#
################################################

classes = ['background', 'brain', 'skull', 'misc']
number_of_classification_labels = len(classes)
image_modalities = ["T1"]
channel_size = len(image_modalities)

unet_model = antspynet.create_unet_model_3d((*template_size, channel_size),
    number_of_outputs=number_of_classification_labels,
    number_of_filters=(16, 32, 64, 128), dropout_rate=0.0,
    convolution_kernel_size=3, deconvolution_kernel_size=2,
    weight_decay=0.0)

weights_filename = scripts_directory + "brainExtraction.h5"
if os.path.exists(weights_filename):
    unet_model.load_weights(weights_filename)

dice_loss = antspynet.multilabel_dice_coefficient(dimensionality=3, smoothing_factor=0.)
surface_loss = antspynet.multilabel_surface_loss()

def multilabel_combined_loss(alpha=0.5):
    def multilabel_combined_loss_fixed(y_true, y_pred):
        loss = (alpha * dice_loss(y_true, y_pred) + 
                (1-alpha) * surface_loss(y_true, y_pred)) 
        return(loss)
    return multilabel_combined_loss_fixed

unet_model.compile(optimizer=tf.keras.optimizers.legacy.Adam(),
                   loss=multilabel_combined_loss(0.75),
                   metrics=dice_loss)

################################################
#
#  Load the brain data
#
################################################

print("Loading braindata.")

data_directory = base_directory + '/Data/'

training_image_files = list()
training_segmentation_files = glob.glob(data_directory + "**/*Mask.nii.gz", recursive=True)

if len(training_segmentation_files) == 0:
    print(data_directory)
    raise ValueError("No training images.")

for i in range(len(training_segmentation_files)):
    image_file = training_segmentation_files[i].replace("Mask.nii.gz", ".nii.gz")
    if not os.path.exists(image_file):
        raise ValueError("Template file doesn't exist.")
    training_image_files.append(image_file)

print("Total training image files: ", len(training_image_files))
print( "Training")

###
#
# Set up the training generator
#

batch_size = 4

generator = batch_generator(batch_size=batch_size,
                    template=template,
                    training_image_files=training_image_files,
                    training_segmentation_files=training_segmentation_files,
                    do_histogram_intensity_warping=True,
                    do_histogram_equalization=True,
                    do_histogram_rank_intensity=True,
                    do_simulate_bias_field=True,
                    do_add_noise=True,
                    do_random_transformation=True,
                    do_random_contralateral_flips=True,
                    do_resampling=True,
                    verbose=False)

track = unet_model.fit(x=generator, epochs=200, verbose=1, steps_per_epoch=32,
    callbacks=[
       keras.callbacks.ModelCheckpoint(weights_filename, monitor="loss",
           save_best_only=True, save_weights_only=True, mode='auto', verbose=1),
       keras.callbacks.ReduceLROnPlateau(monitor='loss', factor=0.75,
          verbose=1, patience=20, mode='auto'),
    #    keras.callbacks.EarlyStopping(monitor='loss', min_delta=0.000001,
    #       patience=20)
       ]
   )

unet_model.save_weights(weights_filename)

File: DktMalfLabeling/Version0/Outer/trainModel.R
library( ANTsR )
library( ANTsRNet )
library( keras )
library( tensorflow )
library( ggplot2 )

keras::backend()$clear_session()

Sys.setenv( "CUDA_VISIBLE_DEVICES" = "1,2" )

baseDirectory <- '/home/ntustison/Data/DktMalfLabeling/'
scriptsDirectory <- paste0( baseDirectory, 'ScriptsOuter/' )
source( paste0( scriptsDirectory, 'batchGenerator.R' ) )

templateDirectory <- '/home/ntustison/Data/DktMalfLabeling/Templates/'
templateX <- antsImageRead( paste0( templateDirectory, "croppedMNI152.nii.gz" ) )
templateSize <- c( 96L, 112L, 96L )

spatialPriorFiles <- list.files( path = paste0( templateDirectory, "croppedMNI152DktPriorLabels/" ),
		    	         pattern = "priorDkt", full.names = TRUE, recursive = FALSE )

spatialPriors <- list()
for( i in seq.int( length( spatialPriorFiles ) ) )
  {
  spatialPriors[[i]] <- resampleImage( antsImageRead( spatialPriorFiles[i] ), templateSize, useVoxels = TRUE, interpType = 0 ) 	 
  }	 
template <- resampleImage( templateX, templateSize, useVoxels = TRUE, interpType = 0 )

################################################
#
#  Create the model and load weights
#
################################################

#classes <- c( 0, 4, 6, 7, 10:18, 24, 26, 28, 30, 43, 44:46, 49:54, 58, 60, 91:92, 
#              630:632, 1002:1003, 1005:1031, 1034:1035, 2002:2003, 2005:2031, 2034:2035 )

classes <- c( 0, 1002:1003, 1005:1031, 1034:1035, 2002:2003, 2005:2031, 2034:2035 )

numberOfClassificationLabels <- length( classes )
imageModalities <- c( "T1" )
channelSize <- length( imageModalities ) + length( spatialPriors )

unetModel <- createUnetModel3D( c( templateSize, channelSize ),
  numberOfOutputs = numberOfClassificationLabels, mode = 'classification',
  numberOfLayers = 4, numberOfFiltersAtBaseLayer = 16, dropoutRate = 0.0,
  convolutionKernelSize = c( 3, 3, 3 ), deconvolutionKernelSize = c( 2, 2, 2 ),
  weightDecay = 1e-5, addAttentionGating = TRUE )

brainWeightsFileName <- paste0( scriptsDirectory, "/dktLabelingOuterWithSpatialPriors.h5" )
if( file.exists( brainWeightsFileName ) )
  {
  load_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )
  # } else {
  # stop( "Weights file doesn't exist.\n" )  
  }

metric_multilabel_dice_coefficient <-
  custom_metric( "multilabel_dice_coefficient",
    multilabel_dice_coefficient )

loss_dice <- function( y_true, y_pred ) {
   -multilabel_dice_coefficient(y_true, y_pred)
   }
attr(loss_dice, "py_function_name") <- "multilabel_dice_coefficient"

# unetModel %>% compile(
#    optimizer = optimizer_adam(),
#   loss = tensorflow::tf$keras$losses$CategoricalCrossentropy(),
#   metrics = c( 'accuracy', metric_multilabel_dice_coefficient, metric_categorical_crossentropy ) )

lossWeights <- rep( 1.0, numberOfClassificationLabels )
lossWeights[1] <- 0.001
weighted_loss <- weighted_categorical_crossentropy( weights = lossWeights )

unetModel %>% compile(
  optimizer = optimizer_adam(),
  loss = loss_dice, # weighted_loss, # categorical_focal_loss( alpha = 0.25, gamma = 2.0 ),
  metrics = c( 'accuracy', metric_multilabel_dice_coefficient, metric_categorical_crossentropy ) )

################################################
#
#  Load the brain data
#
################################################

cat( "Loading brain data.\n" )

brainImages <- c( 
  Sys.glob( "/raid/data_NT/CorticalThicknessData2014/*/ThicknessAnts/*xMniInverseWarped.nii.gz" ),
  Sys.glob( "/home/ntustison/Data/SixTissueSegmentation/Data/*/*xMniInverseWarped.nii.gz" ) )

trainingImageFiles <- c()
trainingSegmentationImageFiles <- c()

pb <- txtProgressBar( min = 0, max = length( brainImages ), style = 3 )
for( i in seq_len( length( brainImages ) ) )
  {
  setTxtProgressBar( pb, i )

  image <- brainImages[i]
  seg <- gsub( "InverseWarped", "RefinedMalfLabels", image )

  if( ! file.exists( image ) || ! file.exists( seg ) )
    {
    next  
    # stop( "Mask doesn't exist." )  
    }

  trainingImageFiles <- append( trainingImageFiles, image )
  trainingSegmentationImageFiles <- append( trainingSegmentationImageFiles, seg )
  }
cat( "\n" )  

cat( "Total training image files: ", length( trainingImageFiles ), "\n" )

cat( "\nTraining\n\n" )

###
#
# Set up the training generator
#

batchSize <- 4L

# Split trainingData into "training" and "validation" componets for
# training the model.

numberOfData <- length( trainingImageFiles )
sampleIndices <- sample( numberOfData )

validationSplit <- floor( 0.8 * numberOfData )
trainingIndices <- sampleIndices[1:validationSplit]
numberOfTrainingData <- length( trainingIndices )
validationIndices <- sampleIndices[( validationSplit + 1 ):numberOfData]
numberOfValidationData <- length( validationIndices )

###
#
# Run training
#

track <- unetModel %>% fit_generator(
  generator = batchGenerator( batchSize = batchSize,
                              imageSize = templateSize, 
                              template = template,
			      spatialPriors = spatialPriors,
                              images = trainingImageFiles[trainingIndices],
                              segmentationImages = trainingSegmentationImageFiles[trainingIndices],
                              segmentationLabels = classes,
                              doRandomContralateralFlips = FALSE,
                              doDataAugmentation = FALSE
                            ),
  steps_per_epoch = 100L,
  epochs = 200L,
  validation_data = batchGenerator( batchSize = batchSize,
                                    imageSize = templateSize, 
                                    template = template,
				    spatialPriors = spatialPriors,
                                    images = trainingImageFiles[validationIndices],
                                    segmentationImages = trainingSegmentationImageFiles[validationIndices],
                                    segmentationLabels = classes,
                                    doRandomContralateralFlips = FALSE,
                                    doDataAugmentation = FALSE
                                  ),  
  validation_steps = 40L,
  callbacks = list(
    callback_model_checkpoint( brainWeightsFileName,
      monitor = 'val_loss', save_best_only = TRUE, save_weights_only = TRUE,
      verbose = 1, mode = 'auto' ), 
     callback_reduce_lr_on_plateau( monitor = 'val_loss', factor = 0.3,
       verbose = 1, patience = 10, mode = 'auto' ),
    callback_early_stopping( monitor = 'val_loss', min_delta = 0.001,
      patience = 20 )
  )
)

save_model_weights_hdf5( unetModel, filepath = brainWeightsFileName )
